solarforecastarbiter/pvmodel.py

```diff
@@ -0,0 +1,428 @@
+            _aoi_tracking,
+            modeling_parameters.axis_tilt,
+            modeling_parameters.axis_azimuth,
+            modeling_parameters.maximum_rotation_angle,
+            modeling_parameters.backtrack,
+            modeling_parameters.ground_coverage_ratio
+        )
+    else:
+        raise TypeError('Invalid modeling_parameters type %s' %
+                        type(modeling_parameters))
+
+
+def _aoi_fixed(surface_tilt, surface_azimuth, apparent_zenith, azimuth):
+    """
+    Calculate AOI for fixed system, bundle return with tilt, azimuth for
+    consistency with similar tracker function.
+
+    Parameters
+    ----------
+    surface_tilt : float
+    surface_azimuth : float
+    apparent_zenith : pd.Series
+        Solar apparent zenith
+    azimuth : pd.Series
+        Solar azimuth
+
+    Returns
+    -------
+    surface_tilt : pd.Series, surface_azimuth : pd.Series, aoi : pd.Series
+    """
+    aoi = pvlib.irradiance.aoi(surface_tilt, surface_azimuth,
+                               apparent_zenith, azimuth)
+    return surface_tilt, surface_azimuth, aoi
+
+
+def _aoi_tracking(axis_tilt, axis_azimuth, maximum_rotation_angle, backtrack,
+                  ground_coverage_ratio, apparent_zenith, azimuth):
+    """
+    Calculate AOI, surface tilt, and surface azimuth for tracking system.
+
+    Parameters
+    ----------
+    axis_tilt : float
+    axis_azimuth : float
+    maximum_rotation_angle : float
+    backtrack : bool
+    ground_coverage_ratio : float
+    apparent_zenith : pd.Series
+        Solar apparent zenith
+    azimuth : pd.Series
+        Solar azimuth
+
+    Returns
+    -------
+    surface_tilt : pd.Series, surface_azimuth : pd.Series, aoi : pd.Series
+    """
+    tracking = pvlib.tracking.singleaxis(
+        apparent_zenith,
+        azimuth,
+        axis_tilt=axis_tilt,
+        axis_azimuth=axis_azimuth,
+        max_angle=maximum_rotation_angle,
+        backtrack=backtrack,
+        gcr=ground_coverage_ratio
+    )
+    surface_tilt = tracking['surface_tilt']
+    surface_azimuth = tracking['surface_azimuth']
+    aoi = tracking['aoi']
+    return surface_tilt, surface_azimuth, aoi
+
+
+def calculate_poa_effective(surface_tilt, surface_azimuth, aoi,
+                            apparent_zenith, azimuth, ghi, dni, dhi):
+    """
+    Calculate effective plane of array irradiance from system metadata,
+    solar position, and irradiance components. Accounts for AOI losses.
+
+    Parameters
+    ----------
+    surface_tilt : float or pd.Series
+    surface_azimuth : float or pd.Series
+    aoi : pd.Series
+    apparent_zenith : pd.Series
+        Solar apparent zenith
+    azimuth : pd.Series
+        Solar azimuth
+    ghi : pd.Series
+    dni : pd.Series
+    dhi : pd.Series
+
+    Returns
+    -------
+    poa_effective : pd.Series
+    """
+    dni_extra = pvlib.irradiance.get_extra_radiation(apparent_zenith.index)
+    poa_sky_diffuse = pvlib.irradiance.haydavies(
```

Comment:
I think you want `pvlib.irradiance.get_total_irradiance` here

Reply:
I considered using `get_total_irradiance`. I thought maybe it wasn't worth it since we have to recalculate POA because of the AOI modifier and we don't need the flexibility of choosing different models. I could go either way.

Reply:
Another option might be to add an optional `aoi_modifier` kwarg to `get_total_irradiance`.

Reply:
The argument list is for `get_total_irradiance` not `haydavies` I thought it was a copy/paste mistake, but now that I look more closely, it doesn't match either.  `dni_extra` is listed twice.  `model` is used by `get_total_irradiance` not `haydavies`.

Reply:
yikes... assuming I can code correctly, do you have a preference for 1 using the ` get_total_irradiance` wrapper + calculating `poa_effective` or 2 explicitly calling each of the component functions here?

Reply:
I am assuming we want to leverage pvlib, except for the class layer.  If so then `get_total_irradiance` is preferred.  Using the base functions is OK, but risks duplicating functions already in pvlib, and I don't see any reason to do that.  Willing to listen.

Reply:
My other reason for preferring `get_total_irradiance` is keep the `model` kwarg as an option for future work.

Reply:
cde6d84 changes the `haydavies` call to `get_sky_diffuse`. I think this is more clear when AOI losses need to be applied to the direct component. It's also slightly more efficient because we already have AOI (`get_total_irradiance` calls `pvlib.irradiance.aoi` internally and its a relatively expensive function). I agree that `get_total_irradiance` is preferred if we don't have AOI losses.